File: typescript/src/domain/repository.ts
import type { CollaborativeDocument, DocumentId } from "./document.js";

export interface DocumentRepository {
  save(document: CollaborativeDocument): Promise<void>;
  findById(id: DocumentId): Promise<CollaborativeDocument | undefined>;
  list(): Promise<readonly CollaborativeDocument[]>;
}

export class InMemoryDocumentRepository implements DocumentRepository {
  private readonly documents = new Map<DocumentId, CollaborativeDocument>();

  async save(document: CollaborativeDocument): Promise<void> {
    this.documents.set(document.id, cloneDocument(document));
  }

  async findById(id: DocumentId): Promise<CollaborativeDocument | undefined> {
    const document = this.documents.get(id);
    return document ? cloneDocument(document) : undefined;
  }

  async list(): Promise<readonly CollaborativeDocument[]> {
    return [...this.documents.values()].map(cloneDocument);
  }
}

function cloneDocument(document: CollaborativeDocument): CollaborativeDocument {
  return { ...document };
}

File: go/internal/documents/adapters/db/repository.go
// Package db provides the in-memory DocumentRepository adapter.
//
// Mirrors typescript/src/documents-service/adapters/db/makeInMemoryDocumentRepository.ts.
package db

import (
	"context"
	"sync"

	"rebac-primer/internal/documents"
)

// InMemoryDocumentRepository is a map-backed [documents.DocumentRepository]
// for tests and demos.  Save stores a struct copy — mutations to the caller's
// value after Save do not affect the stored document (snapshot semantics, just
// like the TypeScript version).
type InMemoryDocumentRepository struct {
	mu   sync.RWMutex
	docs map[string]documents.CollaborativeDocument
}

// New creates an empty repository.
func New() *InMemoryDocumentRepository {
	return &InMemoryDocumentRepository{
		docs: make(map[string]documents.CollaborativeDocument),
	}
}

// Compile-time assertion: *InMemoryDocumentRepository must satisfy DocumentRepository.
var _ documents.DocumentRepository = (*InMemoryDocumentRepository)(nil)

// Save stores or replaces a document (idempotent on ID).
func (r *InMemoryDocumentRepository) Save(_ context.Context, doc documents.CollaborativeDocument) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.docs[doc.ID] = doc // struct copy → snapshot semantics
	return nil
}

// FindByID retrieves a document by ID.  Returns nil (not an error) when not found.
func (r *InMemoryDocumentRepository) FindByID(_ context.Context, id string) (*documents.CollaborativeDocument, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	doc, ok := r.docs[id]
	if !ok {
		return nil, nil
	}
	return &doc, nil
}
